### PR TITLE
[foreach] Fix 0-size handling for real for real

### PR DESCRIPTION
**Pull request by janeyx99**
@crcrpar's last attempt to fix the 0-size problem unfortunately did not pass all cases. See my comment in https://github.com/pytorch/pytorch/issues/100701. When we have a tail tensor of size 0, the old code would mess with the chunk logic to check the previous tensor's length. This is flawed because:
1. if the previous tensor was also 0 sized, (so a tensor list of [tensor, tensor, tensor, ..., 0-sized tensor, 0-sized tensor],) chunks would still be 0 and the nested for loop would be missed.
2. the nested forloop pronounces side effects on tensorListMeta that _shouldn't_ be there! This can mess up the compute in unexpected ways that I haven't really needed to reason through.

We noticed that the problem had not been fixed due to an internal report. This PR solves the issue by:
- removing the finagling of chunks when the tail tensor is 0-sized
- adding a surefire way for the kernel to be launched in the case where the last tensor is 0-sized AND there's content in the metadata, signifying there is stuff to compute still.

## test plan

As I went through the code, I also added some comments explaining what's up and modified our tensor inputs to ensure that this case is tested in the test_parity test in test_foreach.py. Yes, I do realize there is quite a bit of duplication and that this file could be due for a refactor. That said, the primary goal of this PR is to fix the pretty egregious bug and refactoring can be a followup.

cc @awgu @crcrpar 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109402

### CODE REVIEW

**Comment by crcrpar** [comment]
I vaguely remember ngimel suggested it should be possible to filter zero size tensors and pass non zero size tensors to multi tensor apply kernel. I failed to do so that time but would it be worth a try now?

**Comment by janeyx99** [comment]
haha i remember us saying we should refactor. i've taken a look and the refactoring wouldn't get more efficient than now, and could get a bit more complicated due to the fact that blocks OR tensors could fill up. could be worth a shot rewriting with filtering, but running through the tensors once seems better than having two passes.

**Comment by janeyx99** [comment]
actually we may be able to do so as a part of the check fast path api...that may be a lot easier....especially because currently this PR still wouldn't fix foreach_norm.

File: torch/testing/_internal/opinfo/core.py [hunk 1]
```diff
@@ -2637,8 +2637,11 @@ def sample_inputs_foreach(
             for _ in range(N)
         ]
     else:
+        # if N > 5, interweave some empty tensors + have the last tensor be 0-sized to fully fix #100701
```
Comment (janeyx99) [inline torch/testing/_internal/opinfo/core.py:2640]:
```suggestion
        # if N >= 5, interweave some empty tensors + have the last tensor be 0-sized to fully fix #100701
```

**Comment by janeyx99** [comment]
Update: I've gone on a grand endeavor to do filtering first, but it is a much more involved change that seems to affect other parts of the stack. (See my attempt at https://github.com/pytorch/pytorch/pull/109550).

The safest path of procession is:
- tidy up and land this change
- land the big change cautiously as a refactor

here's to hoping for green CI

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -8758,10 +8764,11 @@ def __call__(self, opinfo, device, dtype, requires_grad, **kwargs):
         foreach_inputs_sample_func(1, False, False),
         backward_requires_result=True,
         decorators=(
+            # due to https://github.com/pytorch/pytorch/pull/102427 enabling jiterator for complex
             DecorateInfo(
                 toleranceOverride(
                     {
-                        torch.complex64: tol(atol=1e-05, rtol=1e-05)
+                        torch.complex64: tol(atol=3e-04, rtol=2e-05)
```
Comment (janeyx99) [inline torch/testing/_internal/common_methods_invocations.py:8771]: @parth-desai @peterbell10 hi 3e-04 seems like a decently large atol to me, but I did confirm locally that the reason for this disparity is the jiterator change. One can repo the following provides different results before and after 102427.

```
import torch
x = torch.tensor(-7.8167-0.0451j, device='cuda:0')
torch.set_printoptions(precision=10)
print(torch.tan(x))
print(torch._foreach_tan([x])[0])
print(torch._foreach_tan([x.to(device="cpu")])[0])
```

Before:
image

After:
image

This PR just happened to catch this since I added more sample inputs to test the empty tensor case so the seed changed. I'm wondering if this is acceptable or whether an issue should be raised to call this out.

Comment (janeyx99) [reply torch/testing/_internal/common_methods_invocations.py:8771]: Or one can run python test/test_foreach.py -k test_parity__foreach_tan_slowpath_outplace_cuda_complex64 without the tolerance changes to repro as well

File: aten/src/ATen/native/cuda/ForeachReduceOp.cu [hunk 3]
```diff
@@ -209,10 +209,18 @@ std::vector<Tensor> foreach_tensor_norm_cuda(
...
   result.reserve(ntensors);
-  for (const auto& i : c10::irange(ntensors)) {
-    result.emplace_back(ret_per_tensor[i]);
+  int i = 0;
+  for (const auto& t : tensors) {
+    if (t.numel() != 0) {
+      result.emplace_back(ret_per_tensor[i]);
+      i++;
+    } else {
+      result.emplace_back(at::zeros({}, options));
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachReduceOp.cu:222]: you still need output of correct shape, no?

Comment (janeyx99) [reply aten/src/ATen/native/cuda/ForeachReduceOp.cu:222]: this is the right shape! the only reduce op we have is _foreach_norm and so the output is a 0dim scalar for every tensor.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (peterbell10) [reply torch/testing/_internal/common_methods_invocations.py:8771]: Agreed that does look quite bad. I think it's okay to revert the changes in [UnaryGeometricTanKernel.cu](https://github.com/pytorch/pytorch/pull/102427/files#diff-2d42c32367b57db301b8cc315096655985730f566495580b45bd9ecf3e5a77fe).

Comment (ngimel) [reply torch/testing/_internal/common_methods_invocations.py:8771]: jiterator uses different complex math implementations (from llvm) than thrust (which is used throughout eager codebase), I think we already had similar discrepancies with sigmoid? Worth checking.

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 4]
```diff
@@ -132,6 +132,8 @@ void multi_tensor_apply(
     if (tensor_lists[0][t].numel() == 0) {
       n_zero_tensors++;
       if (t != n_tensors - 1) {
+        // allow launching of the last kernel with accumulated
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:135]: what do you think of the following:
1) remove check on line 134, unconditionally cycle if there's a zero-sized tensor
2) remove block on line 150, you'll never get there
3) remove last_chunk computation on line 167 and last chunk condition on line 169
4) flush remaining blocks on line 197 if loc_block_info is non-zero, regardless of chunks

Does it work?
1) all zero tensors, we never go into meat of the loop, loc_block_info is zero, nothing is launched, good
2) a string of zero element tensors at the end, we never go into loop for them, but loc_block_info is (probably) nonzero. It it's nonzero, we'll flush whatever was accumulated, if it's already zero, we are good.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (janeyx99) [reply torch/testing/_internal/common_methods_invocations.py:8771]: Opened an issue https://github.com/pytorch/pytorch/issues/110014

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh (hunk 4 above)
Comment (janeyx99) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:135]: I did have this as a fleeting thought--now that we have test cases, what you're saying sounds reasonable and wouldn't hurt to try! I'm running the tests now--will let u know how it goes

Comment (janeyx99) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:135]: looks promising!

Comment (janeyx99) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh]: updated PR to be simpler :) thanks @ngimel for the adept suggestions

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -8583,7 +8583,13 @@ def __call__(self, opinfo, device, dtype, requires_grad, **kwargs):
         _foreach_inputs_kwargs["requires_grad"] = requires_grad
         _foreach_inputs_kwargs["zero_size"] = False
 
-        for num_tensors, rightmost_arg_type in itertools.product(num_input_tensors, self._rightmost_arg_types):
+        # add empty tensor interspersion to test fully fixing #100701
+        for num_tensors, rightmost_arg_type, intersperse_empty_tensors in itertools.product(
+                num_input_tensors, self._rightmost_arg_types, (True, False)):
+            if intersperse_empty_tensors and (num_tensors != max(num_input_tensors) or str(device) == 'cpu'):
+                # generate interspersed empty tensors for only 1 N on non-cpu device to lessen redundancy
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:8590]: only 1 N ?

**Review by albanD — APPROVED** [review]
Sounds good!
Let's follow up on the precision issue in details

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (janeyx99) [reply torch/testing/_internal/common_methods_invocations.py:8590]: Yes, like the largest N.
